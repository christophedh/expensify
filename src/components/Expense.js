import React from 'react'
import { connect } from 'react-redux'
import { removeExpense } from '../actions/expense'
import { NavLink } from 'react-router-dom'
const Expense = ({ dispatch, id, description, amount, createdAt }) => (
    <div>
        <h3>
            <NavLink to={`/edit/${id}`} activeClassName="is-active" exact>
                {description}
            </NavLink>
        </h3>
        <p>
            {amount} - {createdAt}
        </p>
        <button
            onClick={() => {
                dispatch(removeExpense({ id }))
            }}
        >
            Remove
        </button>
    </div>
)

export default connect()(Expense)
