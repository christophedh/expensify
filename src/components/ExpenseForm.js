import React from 'react'
import moment from 'moment'
import { addExpense, editExpense } from '../actions/expense'
import { SingleDatePicker } from 'react-dates'
import 'react-dates/lib/css/_datepicker.css'
import { connect } from 'react-redux'
const uuidv4 = require('uuid/v4')

class ExpenseForm extends React.Component {
    state = {
        description: '',
        note: '',
        amount: '',
        createdAt: moment(),
        calendarFocused: false
    }
    componentDidMount() {
        const { expense } = this.props

        if (expense) {
            this.setState(() => ({
                ...expense,
                createdAt: moment(expense.createdAt)
            }))
        }
    }
    onDescriptionChange = e => {
        const description = e.target.value
        this.setState(() => ({ description }))
    }
    onNoteChange = e => {
        const note = e.target.value
        this.setState(() => ({ note }))
    }
    onAmountChange = e => {
        const amount = e.target.value

        if (amount.match(/^\d*(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }
    }
    onDateChange = createdAt => {
        this.setState(() => ({ createdAt }))
    }
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    }
    onSubmit = e => {
        e.preventDefault()
        const { dispatch, expense } = this.props
        const { description, note, amount, createdAt } = this.state
        if (expense) {
            dispatch(
                editExpense({
                    expenseId: expense.id,
                    expenseUpdate: {
                        description,
                        note,
                        amount: parseFloat(amount),
                        createdAt: createdAt.unix()
                    }
                })
            )
        } else {
            dispatch(
                addExpense({
                    id: uuidv4(),
                    description,
                    note,
                    amount: parseFloat(amount),
                    createdAt: createdAt.unix()
                })
            )
        }
        this.props.onSubmitCompleted()
    }
    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        id={uuidv4()}
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    ></textarea>
                    <button>
                        {this.props.expense ? 'Edit Expense' : 'Add Expense'}
                    </button>
                </form>
            </div>
        )
    }
}

export default connect()(ExpenseForm)
