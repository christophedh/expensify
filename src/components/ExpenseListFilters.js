import React from 'react'
import { connect } from 'react-redux'
import { setTextFilter, sortByDate, sortByAmount } from '../actions/filter'
import { DateRangePicker } from 'react-dates'
import { setStartDate, setEndDate } from '../actions/filter'
import moment from 'moment'

export class ExpenseListFilters extends React.Component {
    state = {
        startDate: undefined,
        endDate: undefined,
        focusedInput: undefined
    }
    componentDidMount() {
        const {
            filters: { startDate, endDate }
        } = this.props

        this.setState({
            startDate: moment(startDate),
            endDate: moment(endDate)
        })
    }

    onTextChange = e => {
        const text = e.target.value
        this.props.setTextFilter(text)
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.setState({
            startDate,
            endDate
        })
        this.props.setDates(startDate.valueOf(), endDate.valueOf())
    }

    setSortBy = e => {
        const sortBy = e.target.value

        if (sortBy === 'date') {
            this.props.sortByDate()
        } else if (sortBy === 'amount') {
            this.props.sortByAmount()
        }
    }

    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            className="text-input"
                            type="text"
                            placeholder="Search expense"
                            value={this.props.filters.text}
                            onChange={this.onTextChange}
                        />
                    </div>
                    <div className="input-group__item">
                        <select
                            className="select"
                            value={this.props.filters.sortBy}
                            onChange={this.setSortBy}
                        >
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.state.startDate}
                            startDateId={'f2ea0227-995f-4846-a900-24dbbad9f022'}
                            endDate={this.state.endDate}
                            endDateId={'2ea0227-995f-4846-a900-24dbbad9f023'}
                            isOutsideRange={() => false}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.focusedInput}
                            onFocusChange={focusedInput =>
                                this.setState({ focusedInput })
                            }
                        />
                    </div>
                </div>
                <br />
            </div>
        )
    }
}

const mapStateToProps = state => {
    return {
        filters: state.filters
    }
}

const mapDispatchToProps = dispatch => ({
    setTextFilter: text => dispatch(setTextFilter(text)),
    setDates: (startDate = 0, endDate = 0) => {
        dispatch(setStartDate(startDate))
        dispatch(setEndDate(endDate))
    },
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate())
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)
