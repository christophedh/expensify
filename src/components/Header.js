import React from 'react'
import { connect } from 'react-redux'
import { NavLink } from 'react-router-dom'
import { startLogOut } from '../actions/auth'

export class Header extends React.Component {
    render() {
        return (
            <header>
                <h1>Expensify</h1>
                <ul>
                    <li>
                        <NavLink
                            to="/dashboard"
                            activeClassName="is-active"
                            exact
                        >
                            Home
                        </NavLink>
                    </li>
                    <li>
                        <NavLink to="/add" activeClassName="is-active" exact>
                            Add an expense
                        </NavLink>
                    </li>
                </ul>
                <button onClick={this.props.startLogOut}>Logout</button>
            </header>
        )
    }
}

const mapDispatchToProps = dispatch => ({
    startLogOut: () => dispatch(startLogOut())
})
export default connect(undefined, mapDispatchToProps)(Header)
