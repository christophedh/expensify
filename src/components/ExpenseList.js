import React from 'react';
import { connect } from 'react-redux';
import Expense from './Expense';
import selectExpenses from '../selectors/expensesSelector';

const ExpenseList = (props) => (
  <div>
    <h1>Expense List</h1>
    {props.expenses.map((expense) => {
      return <Expense key={expense.id} {...expense} />;
    })}
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
