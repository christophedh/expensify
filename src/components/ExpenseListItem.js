import React from 'react'
import { connect } from 'react-redux'
import Expense from './Expense'
import selectExpenses from '../selectors/expensesSelector'
import selectExpensesTotal from '../selectors/expensesTotal'
export const ExpenseListItem = props => (
    <div>
        <h1>Expense List</h1>
        {props.expenses.length ? (
            props.expenses.map(expense => {
                return <Expense key={expense.id} {...expense} />
            })
        ) : (
            <p>No expenses</p>
        )}
    </div>
)

const mapStateToProps = state => ({
    expenses: selectExpenses(state.expenses, state.filters)
})

export default connect(mapStateToProps)(ExpenseListItem)
