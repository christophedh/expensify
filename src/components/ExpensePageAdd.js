import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
const AddExpensePage = props => {
    const {
        match: {
            params: { id }
        },
        history
    } = props
    console.log(id)
    return (
        <div>
            <h1>Add an expense</h1>
            <ExpenseForm
                onSubmitCompleted={() => {
                    history.push('/')
                }}
            />
        </div>
    )
}

export default connect()(AddExpensePage)
