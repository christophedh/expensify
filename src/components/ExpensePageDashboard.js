import React from 'react'
import ExpenseListItem from './ExpenseListItem'
import ExpenseListFilters from './ExpenseListFilters'
import ExpenseSummary from './ExpenseSummary'

const ExpenseDashboardPage = () => (
    <div>
        <ExpenseSummary />
        <ExpenseListFilters />
        <ExpenseListItem />
    </div>
)

export default ExpenseDashboardPage
