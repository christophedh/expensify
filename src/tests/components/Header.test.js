import React from 'react'
import { shallow } from 'enzyme'
import { Header } from '../../components/Header'

let wrapper, startLogOut

beforeEach(() => {
    startLogOut = jest.fn()
    wrapper = shallow(<Header startLogOut={startLogOut} />)
})
test('should render Header correctly', () => {
    expect(wrapper).toMatchSnapshot()
})

test('should logOut on button click', () => {
    wrapper.find('button').simulate('click')
    expect(startLogOut).toHaveBeenCalled()
})
