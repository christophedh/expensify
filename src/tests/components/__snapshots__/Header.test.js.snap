// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<header>
  <h1>
    Expensify
  </h1>
  <ul>
    <li>
      <NavLink
        activeClassName="is-active"
        exact={true}
        to="/dashboard"
      >
        Home
      </NavLink>
    </li>
    <li>
      <NavLink
        activeClassName="is-active"
        exact={true}
        to="/add"
      >
        Add an expense
      </NavLink>
    </li>
  </ul>
  <button
    onClick={[MockFunction]}
  >
    Logout
  </button>
</header>
`;
